### PR TITLE
fix #366: stream failure if the topic doesn't exist with Pulsar 2.8.0

### LINKED ISSUE #366: [BUG] Stream failure if the topic doesn't exist with Pulsar 2.8.0
**Describe the bug**
With a Pulsar 2.8.0 server, a Flink stream containing Pulsar sources or sinks will fail at startup if the topic doesn't exist.

**To Reproduce**
Create a stream with : 
- Flink 1.13.1
- Pulsar 2.8.0
- pulsar-flink connector 1.13.1.0 (probably with the 1.13.1.1 too as the concerned code doesn't seem to have changed, but I havent tried)
- with a Pulsar source linked to a non existant topic
Start the stream
It will fail with this error :
`21/06/29 16:20:26 WARN taskmanager.Task: Source: Custom Source (3/8)#0 (63ad2d64523188db366fd4fbf8970612) switched from INITIALIZING to FAILED with failure cause: org.apache.pulsar.client.admin.PulsarAdminException$NotFoundException: Topic not exist`

**Expected behavior**
If the topic doesn't exist, it should be created at the first connection of the source/sink without error.

**Additional context**
In the PulsarMetadataReader class of the connector, the method getTopicPartitionsAll() try to get the partitions number from Pulsar doing that :
`admin.topics().getPartitionedTopicMetadata(topic.getTopic()).partitions;
`
With Pulsar 2.7.2, if the topic doesn't exist, the admin API response is a HTTP 200 with this content : {"partitions":0}
With Pulsar 2.8.0, if the topic doesn't exist, the admin API response is a HTTP 404 with this content : {"reason":"Topic not exist"} and will raise a PulsarAdminException$NotFoundException and then fail the Stream.

The change in Pulsar 2.8.0 : https://github.com/apache/pulsar/pull/10601

We did this as a workaround (just for our tests) in the flink-connector :
```
int partNum = 0;
try {
   partNum = admin.topics().getPartitionedTopicMetadata(topic.getTopic()).partitions;
} catch (PulsarAdminException.NotFoundException ignored) {}
```
but there are probably other impacts in the connector, like in this method of the same class :
```
    public boolean topicExists(String topicName) throws PulsarAdminException {
        int partitionNum = admin.topics().getPartitionedTopicMetadata(topicName).partitions;
        if (partitionNum > 0) {
            return true;
        } else {
            admin.topics().getStats(topicName);
        }
        return true;
    }
```
 this method will probably fail too and never return false

### PR DESCRIPTION
**Pull request by shibd**
#366

### CODE REVIEW

File: pulsar-flink-connector/src/main/java/org/apache/flink/streaming/connectors/pulsar/internal/PulsarMetadataReader.java [hunk 1]
```diff
@@ -357,17 +359,34 @@ public SchemaInfo getPulsarSchema(String topic) {
...
+            } catch (PulsarAdminException.NotFoundException e) {
+                log.info("topic<{}> is not exit, auto create <{}> partition to <{}>", topic.getTopic(), partNum, topic.getTopic());
+                try {
+                    createTopic(topic.getTopic(), partNum);
+                } catch (PulsarAdminException.ConflictException conflictException) {
+                    // multi thread may cause concurrent creation
+                }
+            }
+            // pulsar still has the situation of getting 0 partitions, case: client.newProducer(Schema.STRING).topic(topicName).create()
             if (partNum == 0) {
```
Comment (shibd) [inline pulsar-flink-connector/src/main/java/org/apache/flink/streaming/connectors/pulsar/internal/PulsarMetadataReader.java:384]: handle this problem.

https://github.com/streamnative/pulsar-flink/issues/366#issuecomment-873330849

File: pulsar-flink-connector/src/test/java/org/apache/flink/streaming/connectors/pulsar/FlinkPulsarITest.java [hunk 2]
```diff
@@ -173,7 +173,6 @@ public void testCaseSensitiveReaderConf() throws Exception {
         List<Integer> messages =
                 IntStream.range(0, 50).mapToObj(t -> Integer.valueOf(t)).collect(Collectors.toList());
 
-        sendTypedMessages(tp, SchemaType.INT32, messages, Optional.empty());
```
Comment (jianyun8023) [inline pulsar-flink-connector/src/test/java/org/apache/flink/streaming/connectors/pulsar/FlinkPulsarITest.java:176]: Why do you want to delete here?

**Review by jianyun8023 — COMMENTED** [review]
I think it's great~

File: pulsar-flink-connector/src/test/java/org/apache/flink/streaming/connectors/pulsar/FlinkPulsarITest.java (hunk 2 above)
Comment (shibd) [reply pulsar-flink-connector/src/test/java/org/apache/flink/streaming/connectors/pulsar/FlinkPulsarITest.java:176]: This method default create topic, To test the automatic creation of a topic when it does not exist

**Review by jianyun8023 — APPROVED** [review]

**Pull request merged**